libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java

```diff
@@ -0,0 +1,152 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package androidx.media3.exoplayer.rtsp.reader;
+
+import static androidx.media3.common.util.Assertions.checkArgument;
+import static androidx.media3.common.util.Assertions.checkStateNotNull;
+
+import androidx.media3.common.C;
+import androidx.media3.common.Format;
+import androidx.media3.common.util.Log;
+import androidx.media3.common.util.ParsableByteArray;
+import androidx.media3.common.util.Util;
+import androidx.media3.exoplayer.rtsp.RtpPacket;
+import androidx.media3.exoplayer.rtsp.RtpPayloadFormat;
+import androidx.media3.extractor.ExtractorOutput;
+import androidx.media3.extractor.OpusUtil;
+import androidx.media3.extractor.TrackOutput;
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+
+/**
+ * Parses an OPUS byte stream carried on RTP packets and extracts individual samples. Refer to
+ * RFC7845 for more details.
+ */
+/* package */ final class RtpOpusReader implements RtpPayloadReader {
+  private static final String TAG = "RtpOpusReader";
+
+  private final RtpPayloadFormat payloadFormat;
+  private @MonotonicNonNull TrackOutput trackOutput;
+  private long firstReceivedTimestamp;
+  private long startTimeOffsetUs;
+
+  private final int sampleRate;
+  private int previousSequenceNumber;
+  private boolean foundOpusIDHeader;
+  private boolean foundOpusCommentHeader;
+
+  public RtpOpusReader(RtpPayloadFormat payloadFormat) {
+    this.payloadFormat = payloadFormat;
+    this.firstReceivedTimestamp = C.INDEX_UNSET;
+    this.sampleRate = this.payloadFormat.clockRate;
+    this.previousSequenceNumber = C.INDEX_UNSET;
+    this.foundOpusIDHeader = false;
+    this.foundOpusCommentHeader = false;
+  }
+
+  // RtpPayloadReader implementation.
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, int trackId) {
+    trackOutput = extractorOutput.track(trackId, C.TRACK_TYPE_AUDIO);
+    trackOutput.format(payloadFormat.format);
+  }
+
+  @Override
+  public void onReceivingFirstPacket(long timestamp, int sequenceNumber) {
+    this.firstReceivedTimestamp = timestamp;
+  }
+
+  @Override
+  public void consume(
+      ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker) {
+    checkStateNotNull(trackOutput);
+
+    /* RFC7845 Section 3
+     * +---------+ +----------------+ +--------------------+ +-----
+     * |ID Header| | Comment Header | |Audio Data Packet 1 | | ...
+     * +---------+ +----------------+ +--------------------+ +-----
+     */
+    if (!foundOpusIDHeader) {
+      int currPosition = data.getPosition();
```

Comment:
What about maintaining the position in `isOpusIDHeader`? As if calling `isOpusIDHeader` does not have side effect on data position